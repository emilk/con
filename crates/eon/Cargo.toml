[package]
authors.workspace = true
categories.workspace = true
description = "Use the Eon config format with serde"
edition.workspace = true
homepage.workspace = true
include = ["LICENSE-APACHE", "LICENSE-MIT", "**/*.rs", "Cargo.toml"]
keywords.workspace = true
license.workspace = true
name = "eon"
publish = true
readme = "../README.md"
repository.workspace = true
rust-version.workspace = true
version.workspace = true

[lints]
workspace = true

[package.metadata.docs.rs]
all-features = true
targets = ["x86_64-unknown-linux-gnu", "wasm32-unknown-unknown"]


[features]
default = ["serde"]

## Enable serializing/deserializing Eon values with `serde`.
serde = ["dep:serde"]


[dependencies]
eon_syntax.workspace = true

document-features = { workspace = true, optional = true }
indexmap.workspace = true
serde = { workspace = true, optional = true }
vec1.workspace = true


[dev-dependencies]
divan = "0.1.21"
insta.workspace = true
serde = { workspace = true, features = ["derive"] }

[[bench]]
name = "bench_parse"
harness = false
