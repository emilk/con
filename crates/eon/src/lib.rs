//! # Eon: the human-friendly configuration format
//! To learn the Eon syntax, see <https://github.com/emilk/eon>.
//!
//! ## Usage with `serde`
//! Make sure to enable the `serde` feature for `eon` in your `Cargo.toml` (it's currently enabled by default):
//! ```toml
//! [dependencies]
//! eon = { version = "*", features = ["serde"] }
//! ```
//!
//! Deserialize any value that implements `serde::Deserialize` using
//! [`eon::from_str`].
//!
//! Serialize any value that implements `serde::Serialize` into Eon using [`Self::to_string`]
//!
//! ## Usage with [`Value`]
//! You can also treat an Eon document as a dynamically types [`Value`].
//!
//! Load an Eon document into a [`Value`] using [`Value::from_str`].
//! Serialize a [`Value`] into an Eon string using [`Value::format`].
//!
//! ## Reading/writing comments
//! An Eon document can contain comments, which are NOT part of the [`Value`] type.
//! To load and serialize comments, use the low-level [`eon_syntax`] crate instead.
//!
//! ## Feature flags
#![cfg_attr(feature = "document-features", doc = document_features::document_features!())]
#![warn(missing_docs)] // let's keep eon well-documented

mod token_tree_from_value;
mod value;
mod value_from_token_tree;

#[cfg(feature = "serde")]
mod serde;

pub use {
    crate::value::{Map, Number, Value, Variant},
    eon_syntax::{Error, FormatOptions, Result, reformat},
};

/// Exported external crates used by [`eon`].
pub mod external {
    pub use eon_syntax;
    pub use vec1;
}

#[cfg(feature = "serde")]
pub use self::serde::{SerializationError, from_str, to_string, to_value};
