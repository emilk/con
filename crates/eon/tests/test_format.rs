use eon::{Number, Value};

#[test]
fn test_format() {
    let value = Value::Map(
        [
            ("string".to_owned(), Value::String("a string".to_owned())),
            ("integer".to_owned(), Value::Number(Number::from(42))),
            ("float".to_owned(), Value::Number(Number::from(1.3))),
            ("boolean".to_owned(), Value::Bool(true)),
            ("nothing".to_owned(), Value::Null),
            (
                "list".to_owned(),
                Value::List(vec![
                    Value::String("item1".to_owned()),
                    Value::Number(Number::from(1337)),
                ]),
            ),
            (
                "map".to_owned(),
                Value::Map(
                    [
                        (
                            "key1".to_owned(),
                            Value::String(
                                "a string containing \"quotes\" and a newline: \n".to_owned(),
                            ),
                        ),
                        ("key2".to_owned(), Value::Number(Number::from(42.0))),
                    ]
                    .into_iter()
                    .collect(),
                ),
            ),
        ]
        .into_iter()
        .collect(),
    );
    let formatted = value.format(&Default::default());
    insta::assert_snapshot!(formatted);
}
