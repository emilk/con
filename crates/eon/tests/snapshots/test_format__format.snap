---
source: crates/eon/tests/test_format.rs
expression: formatted
---
string: "a string"
integer: 42
float: 1.3
boolean: true
nothing: null
list: ["item1", 1337]
map: {
	key1: 'a string containing "quotes"'
	key2: 42.0
}
